lib/logcat.js

```diff
@@ -1,118 +1,80 @@
-"use strict";
+import { SubProcess } from 'teen_process';
+import log from './logger';
 
-var spawn = require('child_process').spawn
-  , through = require('through')
-  , EventEmitter = require('events').EventEmitter
-  , util = require('util')
-  , _ = require('lodash')
-  , logger = require('./logger');
-
-var Logcat = function (opts) {
-  EventEmitter.call(this);
-  this.adb = opts.adb;
-  this.debug = opts.debug;
-  this.debugTrace = opts.debugTrace;
-  this.proc = null;
-  this.onLogcatStart = null;
-  this.logcatStarted = false;
-  this.calledBack = false;
-  this.logs = [];
-  this.logsSinceLastRequest = [];
-};
-
-util.inherits(Logcat, EventEmitter);
-
-Logcat.prototype.startCapture = function (cb) {
-  this.onLogcatStart = cb;
-  logger.debug("Starting logcat capture");
-  this.proc = spawn(this.adb.path, this.adb.defaultArgs.concat(['logcat']));
-  this.proc.stdout.setEncoding('utf8');
-  this.proc.stderr.setEncoding('utf8');
-  this.proc.on('error', function (err) {
-    logger.error('Logcat capture failed: ' + err.message);
-    if (!this.calledBack) {
-      this.calledBack = true;
-      cb(err);
-    }
-  }.bind(this));
-  this.proc.on('exit', function (code, signal) {
-    logger.debug('Logcat terminated with code ' + code + ', signal ' + signal);
+class Logcat {
+  constructor (opts = {}) {
+    this.adb = opts.adb;
+    this.debug = opts.debug;
+    this.debugTrace = opts.debugTrace;
     this.proc = null;
-  }.bind(this));
-  this.proc.stdout.pipe(through(this.onStdout.bind(this)));
-  this.proc.stderr.pipe(through(this.onStderr.bind(this)));
-};
-
-Logcat.prototype.stopCapture = function (cb) {
-  logger.debug("Stopping logcat capture");
-  if (this.proc === null) {
-    logger.debug("Logcat already stopped");
-    cb();
-    return;
+    this.logs = [];
+    this.logsSinceLastRequest = [];
   }
-  this.proc.on('exit', function () {
-    cb();
-  });
-  this.proc.kill();
-  this.proc = null;
-};
 
-Logcat.prototype.onStdout = function (data) {
-  this.onOutput(data, '');
-};
-
-Logcat.prototype.onStderr = function (data) {
-  if (/execvp\(\)/.test(data)) {
-    logger.error('Logcat process failed to start');
-    if (!this.calledBack) {
-      this.calledBack = true;
-      this.onLogcatStart(new Error("Logcat process failed to start"));
-      return;
-    }
+  startCapture () {
+    return new Promise(async (resolve, reject) => {
+      log.debug("Starting logcat capture");
+      this.proc = new SubProcess(this.adb.path, this.adb.defaultArgs.concat(['logcat']));
+      await this.proc.start(0);
+      this.proc.on('exit', (code, signal) => {
+        log.info(`Logcat terminated with code ${code}, signal ${signal}`);
```

Comment:
then maybe before we stop it we should remove the exit listener? we're using SubProcess, so when we `await this.proc.stop();` we know that it has stopped after that, so we don't also need to be listening in that case. so what we could do is this to stop:

```js
this.proc.removeAllListeners('exit');
await this.proc.stop();
```